types/redux-first-router/index.d.ts

```diff
@@ -138,70 +138,70 @@ export type Listener = (
+        state: Nullable<TState>
+    ): Nullable<TState>;
     getStateForAction(
         action: object,
-        state: Nullable<object>
-    ): Nullable<object>;
+        state: Nullable<TState>
+    ): Nullable<TState>;
     getPathAndParamsForState(
-        state: object
+        state: TState
     ): { path: Nullable<string>; params: Nullable<Params> };
     getActionForPathAndParams(path: string): Nullable<object>;
 }
 
-export interface Navigator {
-    router: Router;
+export interface Navigator<TState = any> {
+    router: Router<TState>;
 }
 
-export interface Navigators {
-    [key: string]: Navigator;
+export interface Navigators<TState = any> {
+    [key: string]: Navigator<TState>;
 }
 
-export type SelectLocationState = (state: object) => LocationState;
-export type SelectTitleState = (state: object) => string;
+export type SelectLocationState<TKeys = {}, TState = any> = (state: TState) => LocationState<TKeys, TState>;
+export type SelectTitleState<TState = any> = (state: TState) => string;
 
 export interface QuerySerializer {
     stringify(params: Params): string;
     parse(queryString: string): object;
 }
 
-export interface NavigatorsConfig {
-    navigators: Navigators;
-    patchNavigators(navigators: Navigators): void;
+export interface NavigatorsConfig<TKeys = {}, TState = any> {
+    navigators: Navigators<TState>;
+    patchNavigators(navigators: Navigators<TState>): void;
 
     actionToNavigation(
-        navigators: Navigators,
+        navigators: Navigators<TState>,
         action: object, // TODO check this
         navigationAction: Nullable<NavigationAction>,
-        route: Nullable<Route>
+        route: Nullable<Route<TKeys, TState>>
     ): object;
     navigationToAction(
-        navigators: Navigators,
-        store: Store<any>,
-        routesMap: RoutesMap,
+        navigators: Navigators<TState>,
+        store: Store<TState>,
+        routesMap: RoutesMap<TKeys, TState>,
         action: object
     ): {
-        action: object;
-        navigationAction: Nullable<NavigationAction>;
-    };
+            action: object;
+            navigationAction: Nullable<NavigationAction>;
+        };
 }
 
-export interface Options {
-    title?: string | SelectTitleState;
-    location?: string | SelectLocationState;
+export interface Options<TKeys = {}, TState = any> {
+    title?: string | SelectTitleState<TState>;
+    location?: string | SelectLocationState<TKeys, TState>;
     notFoundPath?: string;
     scrollTop?: boolean;
-    onBeforeChange?(dispatch: Dispatch<any>, getState: StateGetter): void;
-    onAfterChange?(dispatch: Dispatch<any>, getState: StateGetter): void;
-    onBackNext?(dispatch: Dispatch<any>, getState: StateGetter): void;
+    onBeforeChange?(dispatch: Dispatch<TState>, getState: StateGetter<TState>): void;
+    onAfterChange?(dispatch: Dispatch<TState>, getState: StateGetter<TState>): void;
+    onBackNext?(dispatch: Dispatch<TState>, getState: StateGetter<TState>): void;
```

Comment:
onBeforeChange, onAfterChange, onBackNext should be changed also.

I you would like to have a typed action you could do:
onBeforeChange<TAction= any>?(dispatch: Dispatch<TAction>, getState: StateGetter<TState>): void;

Reply:
As I mentioned in another comment, this fits the `redux@v4.0.0-beta.1` typings, but current typings for redux expects dispatch to pass a state `S` as type-paramter rather than an action. Would you like the typing to be
```
onBeforeChange?(dispatch: Dispatch<any>, getState: StateGetter<TState>): void;
```
That way it works with the current redux `3.7.2` version aswell as `4.0-beta.1`.

Reply:
Also, your suggestion of having the `onBeforeChange` function take a type parameter
```typescript
onBeforeChange<TAction= any>?(dispatch: Dispatch<TAction>, getState: StateGetter<TState>): void;
```
Doesn't help when you're specifying the callback in `redux@4.0` either. It helps you if you try to call the function on an options object as it gives you an optional type parameter, but when specifying the function and trying to call `dispatch()` it will error with 
```
[ts] Argument of type '{ type: string; payload: {}; }' is not assignable to parameter of type '
```
But as far as I understand there is no way for us to provide that type-paramter when specifying the callback.

Reply:
I would change it back to this indeed.